Use partial order to support different split behaviors

erik-stripe opened the pull request:
This PR moves brushfire-core over to use `spire.algebra.PartialOrder` and `spire.algebra.Order` where applicable. At this point we don't actually _use_ any partial orderings, but the support is there.

Before we merge this I'd like to add some tests to prove this is working, and maybe flesh out support for the _is present_ use-case a bit. What do you all think we need?

Review by @tixxit and @avibryant.

erik-stripe committed:
Use spire.algebra.PartialOrder in Predicate.

This is the minimal commit necessary to switch from using
scala.math.Ordering to spire.algebra.PartialOrder in Predicate.
This leaves other uses of scala.math.Ordering alone -- in fact
we may want to switch those for spire.algebra.Order as well.

erik-stripe committed:
Use spire.algebra.Order where appropriate in brushfire-core.

This commit completes the previous work adding PartialOrder.
It also removes a bunch of unnecessary ordering constraints
we had added (and instances we were capturing). It should
not change how Brushfire actually works (yet).

erik-stripe committed:
Merge remote-tracking branch 'upstream/master' into erik-partial-order

tixxit commented:
Tests are good, but I don't think we need to flesh out the is-present stuff in this PR. Just moving `Dispatched` over to using a `PartialOrder` would be enough.

striation commented:
OK, that makes sense. I'll do that next.

erik-stripe committed:
Move Dispatched over to PartialOrder.

This is an improvement since we have a principled way to
handle the "error case" now.

striation commented:
Alright, `Dispatched` is moved over (that was easy!).

erik-stripe committed:
Add some return types to Dispatched.

erik-stripe committed:
Merge remote-tracking branch 'upstream/master' into erik-partial-order

striation commented:
I updated this to master, should be able to be merged.

?r @tixxit @johnynek

johnynek commented:
This looks good to me.

I do think if we are going to use spire more (and we should) we will get mileage out of the algebra work to make algebird, scalding and spire work better together.

That or give up and make a `spire-algebird` package that gives conversions between the various typeclasses.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/stripe/brushfire?pullRequest=87) <br/>All committers have signed the CLA.
